### PR TITLE
F0 adc wwdg

### PR DESCRIPTION
**Pull request by jaxxzer**
Small things encountered during development.

F3 may need adjustment, too. I can check/test that soon.

### CODE REVIEW

**Comment by jaxxzer** [comment]
> F3 may need adjustment, too. I can check/test that soon.

it looks like f3 needs some work too. checking that now.

File: lib/stm32/f0/adc.c [hunk 1]
```diff
@@ -241,7 +241,7 @@ bool adc_get_watchdog_flag(uint32_t adc)
 
 void adc_clear_watchdog_flag(uint32_t adc)
 {
-	ADC_ISR(adc) = ADC_ISR_AWD1;
+	ADC_ISR(adc) |= ADC_ISR_AWD1;
```
Comment (grevaillot) [inline lib/stm32/f0/adc.c]: According to F0 Reference manual, bits in ADC_ISR register are cleared by writing 1 to them.
By or-ing ADC_ISR with AWD1 bit your code will actually clear all set ISR bits (and awd1) in ISR register, which is not what you want.

File: include/libopencm3/stm32/f0/adc.h [hunk 2]
```diff
@@ -190,8 +190,8 @@ void adc_calibrate_wait_finish(uint32_t adc)
 void adc_enable_analog_watchdog_on_all_channels(uint32_t adc);
 void adc_enable_analog_watchdog_on_selected_channel(uint32_t adc, uint8_t chan);
 void adc_disable_analog_watchdog(uint32_t adc);
-void adc_set_watchdog_high_threshold(uint32_t adc, uint8_t threshold);
-void adc_set_watchdog_low_threshold(uint32_t adc, uint8_t threshold);
+void adc_set_watchdog_high_threshold(uint32_t adc, uint16_t threshold);
+void adc_set_watchdog_low_threshold(uint32_t adc, uint16_t threshold);
```
Comment (grevaillot) [inline include/libopencm3/stm32/f0/adc.h]: +1 - and indeed, same stuff for F3 ADC_TR1..

File: lib/stm32/f0/adc.c [hunk 3]
```diff
@@ -519,9 +519,9 @@ void adc_disable_analog_watchdog(uint32_t adc)
  * @param[in] threshold Unsigned int8. Upper threshold value
```
Comment (grevaillot) [inline lib/stm32/f0/adc.c]: You need to update doxygen threshold type parameter to uint16_t as well (or preferably get rid of parameter type in documentation) - in any case threshold is not int8 anymore.

File: lib/stm32/f0/adc.c [hunk 4]
```diff
@@ -531,9 +531,9 @@ void adc_set_watchdog_high_threshold(uint32_t adc, uint8_t threshold)
  * @param[in] threshold Unsigned int8. Lower threshold value
```
Comment (grevaillot) [inline lib/stm32/f0/adc.c]: here too :)

File: lib/stm32/f0/adc.c (hunk 1 above)
Comment (jaxxzer) [reply lib/stm32/f0/adc.c]: :eyes:  I'll try to deal with this revelation and update it this weekend.

File: lib/stm32/f0/adc.c (hunk 3 above)
Comment (jaxxzer) [reply lib/stm32/f0/adc.c]: done

File: lib/stm32/f0/adc.c (hunk 4 above)
Comment (jaxxzer) [reply lib/stm32/f0/adc.c]: done

File: include/libopencm3/stm32/f0/adc.h (hunk 2 above)
Comment (jaxxzer) [reply include/libopencm3/stm32/f0/adc.h]: done

File: lib/stm32/f0/adc.c (hunk 1 above)
Comment (jaxxzer) [reply lib/stm32/f0/adc.c]: i ditched the commit.

**Comment by jaxxzer** [comment]
wip so I can test this and review my own patch again.

**Comment by jaxxzer** [comment]
Everything is working ok over here: https://github.com/jaxxzer/stm32-boilerplate/blob/adc-wwdg/src/example/adc-wwdg/main.c#L21

tested with nucleo f072 and f334 for so far.

**Comment by karlp** [comment]
Could you elaborate on "unify behavior with f0 adc_enable_analog_watchdog_on_selected_channel" ?  What's the change there? is just using the same code style for both? or was there an error in the past too?

**Comment by jaxxzer** [comment]
updated/elaborated; new commits with better organization and message bodies.

File: include/libopencm3/stm32/common/adc_common_v2.h [hunk 5]
```diff
@@ -193,7 +193,7 @@ specific memorymap.h header before including this header file.*/
 #define ADC_TR1_HT_MASK			0xFFF
 #define ADC_TR1_HT			(0xFFF << ADC_TR1_HT_SHIFT)
 /** TR1_HT: analog watchdog 1 threshold high */
-#define ADC_TR1_HT_VAL(x)		((x) << ADC_TR1_HT_SHIFT)
+#define ADC_TR1_HT_VAL(x)		(((x) & ADC_TR1_HT_MASK) << ADC_TR1_HT_SHIFT)
```
Comment (karlp) [inline include/libopencm3/stm32/common/adc_common_v2.h]: You should fix this for _LT_ as well.

Comment (jaxxzer) [reply include/libopencm3/stm32/common/adc_common_v2.h]: done

**Comment by jaxxzer** [comment]
ping I think this is ready

File: lib/stm32/f3/adc.c [hunk 6]
```diff
@@ -419,37 +416,40 @@ void adc_start_conversion_injected(uint32_t adc)
...
 	reg32 |= (threshold << 16);
-	reg32 &= ~0xff00ffff; /* Clear all bits above 8. */
-	ADC_TR1(adc) = reg32;
-	ADC_TR2(adc) = reg32;
-	ADC_TR3(adc) = reg32;
+	reg32 &= ~mask; /* clear masked bits. */
+
+	ADC_TR1(adc) = (ADC_TR1(adc) & mask) | reg32;
+	ADC_TR2(adc) = (ADC_TR2(adc) & mask) | reg32;
+	ADC_TR3(adc) = (ADC_TR2(adc) & mask) | reg32;
```
Comment (grevaillot) [inline lib/stm32/f3/adc.c]: hmm, s/ADC_TR2/ADC_TR3/

File: lib/stm32/f3/adc.c [hunk 7]
```diff
@@ -419,37 +416,40 @@ void adc_start_conversion_injected(uint32_t adc)
...
 	reg32 = (uint32_t)threshold;
-	reg32 &= ~0xffffff00; /* Clear all bits above 8. */
-	ADC_TR1(adc) = reg32;
-	ADC_TR2(adc) = reg32;
-	ADC_TR3(adc) = reg32;
+	reg32 &= ~mask; /* clear masked bits. */
+
+	ADC_TR1(adc) = (ADC_TR1(adc) & mask) | reg32;
+	ADC_TR2(adc) = (ADC_TR2(adc) & mask) | reg32;
+	ADC_TR3(adc) = (ADC_TR2(adc) & mask) | reg32;
```
Comment (grevaillot) [inline lib/stm32/f3/adc.c]: same here

File: lib/stm32/f3/adc.c (hunk 6 above)
Comment (jaxxzer) [reply lib/stm32/f3/adc.c]: fixed

File: lib/stm32/f3/adc.c (hunk 7 above)
Comment (jaxxzer) [reply lib/stm32/f3/adc.c]: fixed

File: lib/stm32/f3/adc.c [hunk 8]
```diff
@@ -282,14 +282,11 @@ void adc_enable_analog_watchdog_on_all_channels(uint32_t adc)
...
+	ADC_CFGR1(adc) = (ADC_CFGR1(adc) & ~ADC_CFGR1_AWD1CH) |
+			  ADC_CFGR1_AWD1CH_VAL(channel);
 
-	reg32 = (ADC_CFGR1(adc) & ~ADC_CFGR1_AWD1CH); /* Clear bit [4:0]. */
-	if (channel < 18) {
-		reg32 |= channel;
-	}
-	ADC_CFGR1(adc) = reg32;
-	ADC_CFGR1(adc) |= ADC_CFGR1_AWD1SGL;
+	// enable watchdog on a single channel (as opposed to all channels)
```
Comment (karlp) [inline lib/stm32/f3/adc.c]: no c++ comments please.  Further, I don't get this one, it says "unify with f0" but it's still in the f3 code?  is this now the same function in two files?  Is it now just a function that should go into adc_common_v2?

**Comment by karlp** [comment]
I'm still curious about the "stm32f3: unify implementation with f0 adc_enable_analog_watchdog_on_s" commit, but I can pick the other three if you like?

File: lib/stm32/f3/adc.c (hunk 8 above)
Comment (karlp) [reply lib/stm32/f3/adc.c]: Right, I see that that is the case.  I've dropped the c++ comment, and am merging this, and will file a "easy pickings" issue for someone who wants to extract this code, there's no need for it to block this commit

**Comment by karlp** [comment]
Thanks for your patience with this.  I've merged it all in 6fc1ff22..e2ac1a63

**Comment by jaxxzer** [comment]
thanks dawg I'll have a look at the fixups you mentioned

**Pull request closed without merging**